src/analyses/cfg_dominators.h

```diff
@@ -142,96 +296,197 @@ void cfg_dominators_templatet<P, T, post_dom>::initialise(P &program)
+}
+
 /// Computes the MOP for the dominator analysis
 template <class P, class T, bool post_dom>
 void cfg_dominators_templatet<P, T, post_dom>::fixedpoint(P &program)
 {
-  std::list<T> worklist;
-
   if(cfgt::nodes_empty(program))
     return;
 
+  // Initialise entry node to be its own dominator:
   if(post_dom)
     entry_node = cfgt::get_last_node(program);
   else
     entry_node = cfgt::get_first_node(program);
   typename cfgt::nodet &n = cfg.get_node(entry_node);
-  n.dominators.insert(entry_node);
+  const auto entry_node_index = cfg.get_node_index(entry_node);
+  n.dominator = entry_node_index;
 
-  for(typename cfgt::edgest::const_iterator
-      s_it=(post_dom?n.in:n.out).begin();
-      s_it!=(post_dom?n.in:n.out).end();
-      ++s_it)
-    worklist.push_back(cfg[s_it->first].PC);
+  // Calculate a post-ordering on the CFG:
+  assign_postordering(entry_node_index);
 
-  while(!worklist.empty())
-  {
-    // get node from worklist
-    T current=worklist.front();
-    worklist.pop_front();
+  const auto visit_order =
+    get_reverse_postordered_instructions(entry_node_index);
 
-    bool changed=false;
-    typename cfgt::nodet &node = cfg.get_node(current);
-    if(node.dominators.empty())
-    {
-      for(const auto &edge : (post_dom ? node.out : node.in))
-        if(!cfg[edge.first].dominators.empty())
-        {
-          node.dominators=cfg[edge.first].dominators;
-          node.dominators.insert(current);
-          changed=true;
-        }
-    }
+  bool any_change = true;
 
-    // compute intersection of predecessors
-    for(const auto &edge : (post_dom ? node.out : node.in))
+  while(any_change)
+  {
+    any_change = false;
+    for(const auto current : visit_order)
     {
-      const target_sett &other=cfg[edge.first].dominators;
-      if(other.empty())
-        continue;
-
-      typename target_sett::const_iterator n_it=node.dominators.begin();
-      typename target_sett::const_iterator o_it=other.begin();
+      auto &node = cfg.get_node(current);
+      const cfg_nodet *dominator_node = nullptr;
 
-      // in-place intersection. not safe to use set_intersect
-      while(n_it!=node.dominators.end() && o_it!=other.end())
+      // compute intersection of predecessors
+      auto &edges = (post_dom ? node.out : node.in);
+      if(edges.size() != 0)
       {
-        if(*n_it==current)
-          ++n_it;
-        else if(*n_it<*o_it)
-        {
-          changed=true;
-          node.dominators.erase(n_it++);
-        }
-        else if(*o_it<*n_it)
-          ++o_it;
-        else
+        for(const auto &edge : edges)
         {
-          ++n_it;
-          ++o_it;
+          const typename cfgt::nodet &other = cfg[edge.first];
+          if(!other.dominator)
+            continue;
+
+          if(!dominator_node)
+            dominator_node = &other;
+          else
+            dominator_node = &intersect(*dominator_node, other);
         }
       }
 
-      while(n_it!=node.dominators.end())
+      if(
```

Comment:
Maybe preface this line with a comment like "if node.dominator != dominator_node, set node.dominator to dominator_node". It isn't immediately obvious that's why you're comparing the postorder_index.